From 30d5b9d2ff6e7ed13ea47abfd84c20899bb34a36 Mon Sep 17 00:00:00 2001
From: Andi Gerl <andi.gerl@exacttechnology.com>
Date: Wed, 11 Dec 2024 16:29:27 -0500
Subject: [PATCH] uart: nrf: power cycle UART0 after suspending

This solves an issue with increased sleep current.

Also fully re-initialize UART0 on resume. Peripheral needs to be re-initialized after
power cycling.
---
 drivers/serial/uart_nrfx_uarte.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/serial/uart_nrfx_uarte.c b/drivers/serial/uart_nrfx_uarte.c
index d2b36017051..52444887270 100644
--- a/drivers/serial/uart_nrfx_uarte.c
+++ b/drivers/serial/uart_nrfx_uarte.c
@@ -29,11 +29,13 @@ LOG_MODULE_REGISTER(uart_nrfx_uarte, CONFIG_UART_LOG_LEVEL);
 #include <nrfx_ppi.h>
 #define gppi_channel_t nrf_ppi_channel_t
 #define gppi_channel_alloc nrfx_ppi_channel_alloc
+#define gppi_channel_free nrfx_ppi_channel_free
 #define gppi_channel_enable nrfx_ppi_channel_enable
 #elif defined(DPPI_PRESENT)
 #include <nrfx_dppi.h>
 #define gppi_channel_t uint8_t
 #define gppi_channel_alloc nrfx_dppi_channel_alloc
+#define gppi_channel_free nrfx_dppi_channel_free
 #define gppi_channel_enable nrfx_dppi_channel_enable
 #else
 #error "No PPI or DPPI"
@@ -1877,6 +1879,10 @@ static int uarte_nrfx_pm_action(const struct device *dev,
 
 	switch (action) {
 	case PM_DEVICE_ACTION_RESUME:
+		if (cfg->uarte_regs == NRF_UARTE0) {
+			return uarte_instance_init(dev,
+						   IS_ENABLED(CONFIG_UART_0_INTERRUPT_DRIVEN));
+		}
 		if (cfg->flags & UARTE_CFG_FLAG_GPIO_MGMT) {
 			ret = pinctrl_apply_state(cfg->pcfg,
 						  PINCTRL_STATE_DEFAULT);
@@ -1959,6 +1965,13 @@ static int uarte_nrfx_pm_action(const struct device *dev,
 			}
 		}
 
+		/* Work around for issue with increased sleep current on UART0 */
+		if (cfg->uarte_regs == NRF_UARTE0) {
+			*(volatile uint32_t *)0x40002FFC = 0;
+			*(volatile uint32_t *)0x40002FFC;
+			*(volatile uint32_t *)0x40002FFC = 1;
+			gppi_channel_free(data->ppi_ch_endtx);
+		}
 		break;
 	default:
 		return -ENOTSUP;
-- 
2.47.1

